src/Strings.h

```diff
@@ -0,0 +1,388 @@
+                       IteratorT const &end,
+                       RegexT const &regexToken,
+                       std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    std::sregex_iterator iter(begin, end, regexToken);
+    std::sregex_iterator iterEnd;
+
+    while (iter != iterEnd) {
+        std::string const iterStringConst = iter->str();
+        IteratorT beginRef = iterStringConst.begin();
+        IteratorT endRef = iterStringConst.end();
+        callback(beginRef, endRef);
+        ++iter;
+    }
+}
+
+template <typename IteratorT>
+inline std::string TrimAndReplace(IteratorT begin,
+                                  IteratorT end) {
+    //replace all punctuations with spaces
+    std::transform(begin, end, begin, [](char c) {
+        if (std::ispunct(c) )
+            return ' ';
+        return c;
+    });
+
+    std::string result(begin, end);
+    //remove duplicate space
+    result.erase(
+        std::unique(
+            result.begin(),
+            result.end(),
+            [=](char lhs, char rhs){
+                return (lhs == rhs) && (lhs == ' ');
+            }
+        ),
+        result.end()
+    );
+
+    return result;
+}
+
+template <typename IteratorT>
+inline void ParseNgramCharHelper(IteratorT const &begin,
+                                 IteratorT const &end,
+                                 size_t const &ngramRangeMin,
+                                 size_t const &ngramRangeMax,
+                                 std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    std::vector<IteratorT> wordIterVector(IteratorVectorGenerator(begin, end));
+
+    for (size_t strIterOffset = 0; strIterOffset < wordIterVector.size(); ++strIterOffset)  {
+        for (size_t ngramRangeVal = ngramRangeMin; ngramRangeVal <= ngramRangeMax; ++ngramRangeVal) {
+            if (strIterOffset + ngramRangeVal < wordIterVector.size())
+                callback(wordIterVector[strIterOffset], wordIterVector[strIterOffset + ngramRangeVal]);
+        }
+    }
+}
+
+template <
+    typename IteratorT,
+    typename IsDelimiterT
+>
+inline void ParseNgramWord(std::string &input,
+                           IsDelimiterT const &isDelimiter,
+                           size_t const &ngramRangeMin,
+                           size_t const &ngramRangeMax,
+                           std::function<void (IteratorT&, IteratorT&)> const &callback) {
+
+    std::string trimedString(TrimAndReplace(input.begin(), input.end()));
+
+    //wordIterPairVector is used to store the begin and end iterator of words in input
+    std::vector<std::pair<IteratorT, IteratorT>> wordIterPairVector;
+    Details::Parse<IteratorT, IsDelimiterT>(
+        trimedString.begin(),
+        trimedString.end(),
+        isDelimiter,
+        [&wordIterPairVector] (IteratorT &iterBegin, IteratorT &iterEnd) {
+            wordIterPairVector.emplace_back(std::make_pair(iterBegin, iterEnd));
```

Comment:
>emplace_back [](start = 31, length = 12)

This syntax makes no improvement over push_back. On the other hand, emplace_back(iterBegin, iterEnd) makes it more efficient.